lib/pure/asyncdispatch.nim

```diff
@@ -355,30 +355,34 @@ proc `or`*[T, Y](fut1: Future[T], fut2: Future[Y]): Future[void] =
   fut2.callback = cb
   return retFuture
 
-proc all*[T](futs: varargs[Future[T]]): Future[seq[T]] =
+proc all*[T](futs: varargs[Future[T]]): auto =
   ## Returns a future which will complete once all futures in ``futs``
   ## complete.
   ##
   ## The resulting future will hold the values of all awaited futures,
   ## in the order they are passed.
 
-  var
-    retFuture = newFuture[seq[T]]("asyncdispatch.all")
-    retValues = newSeq[T](len(futs))
-    completedFutures = 0
+  when T is void:
+    return foldl(futs, a and b)
```

Comment:
btw, folding here is not as efficient because you're allocating n-1 new futures instead of 1 like you do it in the non-void branch.